### DISCUSSION TITLE
Trying CM3 again! New install

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#1155] [Q&A]
> do-cm3-front.sh                                                         <2>
> /home/dnormandin/TARS/cm3-dist-AMD64_LINUX-d5.11.4/scripts/pkgmap.sh -c "cm3 -build -override $RARGS -DROOT='/home/dnormandin/TARS/cm3-dist-AMD64_LINUX-d5.11.4' " m3core libm3 set sysutils m3middle m3objfile m3linker m3back m3cggen m3quake m3front m3cgcat mklib cm3 m3cc
> package m3cc not found
>  *** cannot find package m3cc / 
I've also installed:
cm3-all-AMD64_LINUX-d5.11.3-Linux5.4.0-20211008

what have I missed? tia ...
_

**Comment by VictorMiasnikov** [comment]
I've public some releases.

https://github.com/VictorMiasnikov/cm3/releases

https://github.com/VictorMiasnikov/cm3/releases/tag/d5.11.9-ZZYYXX-20221108_14-05

 I hope this file:

Schutz_2022-06-08_and_cm3-all-AMD64_LINUX-d5.11.9-AMD64_LINUX(Debian_8.11.1_and_{{GCC_4.9.2}})-2022-11-08_14-05__FIXed__by__VVM__2022-11-14_08-16__AMD64_LINUX__Target__cm3-fbcb95d.tar.xz

can be relative usefull.

**Reply by dukester** [reply]
I DLed your d5.11.9-ZZYYXX-20221108_14-05 - No Joy! I get the same error msg about m3cc compiler.

**Reply by VictorMiasnikov** [reply]
> } I get the same error msg

What is a goal of running do-cm3-front.sh script? 

Distributive(s) named cm3-all-* already contain maximum number of libraries and binary files.

You want recreating cm3 distributive from bootstrap?

**Reply by dukester** [reply]
> > } I get the same error msg
> 
> What is a goal of running do-cm3-front.sh script?

I want to install M3 compiler and linker in order to learn M3. I do not need a GUI front-end.
I want something like gcc & linker. I use emacs for text editor. All simple. All command line.
Can this be done **easily** with CM3? Thanks.

**Reply by VictorMiasnikov** [reply]
> I want to install M3 compiler and ( . . .) in order to learn M3. 

`0) ` Please use as "source code" part of this commit
https://github.com/VictorMiasnikov/cm3/archive/refs/tags/d5.11.9-ZZYYXX-20221108_14-05.tar.gz

  Unpack it to folder
/usr/local/cm3/cm3/

`1)`
       Download this file
https://github.com/VictorMiasnikov/cm3/releases/download/d5.11.9-ZZYYXX-20221108_14-05/cm3-boot-AMD64_LINUX-d5.11.9-20221109_12-50.tar.xz

unpack to any folder.

Run `make`
 
 `2)`
 
 Place resulted file `cm3` to folder  `/usr/local/cm3/bin`

 `3)`   Run:

```
/# cd /usr/local/cm3/cm3/scripts/python
/usr/local/cm3/cm3/scripts/python# export PATH=/usr/local/cm3/bin:$PATH
/usr/local/cm3/cm3/scripts/python# ./boot2min.py c
/usr/local/cm3/cm3/scripts/python# python do-cm3-all.py buildship
```
|
P.S.

> } Excellent! Thank you very much.

You are welcome!

**Reply by dukester** [reply]
Excellent! Thank you very much.

**Reply by dukester** [reply]
/usr/local/cm3/cm3/scripts/python
sudo ./boot2min.py c
fatal error environment variable CM3_INSTALL not set AND cm3 not found in PATH; please fix

The PATH is set!! CM3_INSTALL is not.

**Reply by VictorMiasnikov** [reply]
} fatal error environment variable 

It's strange...
 
This step 
VVM } place cm3 to folder /usr/local/cm3/bin

is done?

"cm3 --version" type info?

( possible dublicate)

**Reply by VictorMiasnikov** [reply]
See later some info

(

>
>  ( Sent 22.02.2023, 20:50)
>
>  I put it into my ~/.zshrc file
>
> which cm3
> /usr/local/cm3/bin/cm3

It looks like Phyton scripts "don't understand" this 

)

**Comment by VictorMiasnikov** [comment]
Alternative method:

Unpack cm3-all-* to /usr/local/cm3.

Unpack source code  to folder
/usr/local/cm3/cm3/

P.S.

 This:
export PATH=/usr/local/cm3/bin:$PATH

steal useful.

**Reply by dukester** [reply]
I put it into my ~/.zshrc file

which cm3
/usr/local/cm3/bin/cm3

**Reply by VictorMiasnikov** [reply]
>
>  ( Sent 22.02.2023, 20:50)
>
>  I put it into my ~/.zshrc file
>
> which cm3
> /usr/local/cm3/bin/cm3

It looks like Phyton scripts "don't understand" this 

( possible dublicated message)

**Comment by dukester** [comment]
I've got a successful install using: [this page](https://github.com/modula3/cm3/wiki/Getting-Started)

**Reply by VictorMiasnikov** [reply]
> I've got a successful install using: [this page  ( VVM: wiki/Getting-Started) ](https://github.com/modula3/cm3/wiki/Getting-Started)

Good news!

**Comment by dukester** [comment]
`MODULE Main;

IMPORT IO;       (* So we can print things *)

VAR name: TEXT;  (* a string variable called "name" *)

BEGIN
  IO.Put("Gimme your name: ");
  name:=IO.GetLine();
  IO.Put("Your name is: " & name & "\n");
END Main.`

> cm3                     
> --- building in AMD64_LINUX ---
> new source -> compiling Main.m3
> "../src/Main.m3", line 9: warning: potentially unhandled exception: IO.Error
> 1 warning encountered
> ****  PARALLEL BACK-END BUILD, M3_PARALLEL_BACK = 20
> new "Main_m.o" -> linking firstProg

It compiles and runs fine. What's the warning about then? Thx

**Reply by VictorMiasnikov** [reply]
Please see about `<*FATAL ANY*>` inside https://github.com/modula3/cm3/discussions/791#discussioncomment-1894374
  and near.
  
  P.S. I continue re-reading and see good example ( I add one line):
  
```
MODULE hello2 EXPORTS Main;
IMPORT IO;
<* FATAL IO.Error *>
<*FATAL ANY*>
VAR name: TEXT;
BEGIN
  TRY
    IO.Put("Gimme your name: ");
    name := IO.GetLine();
  EXCEPT IO.Error => (*do nothing*)
  END;
IO.Put("Your name was entered as: " & name & "\n");
END hello2.
(* HELP  - where do I put
   RAISES {IO.Error}
*)
```

**Comment by RodneyBates** [comment]
On 2/23/23 03:12, VictorMiasnikov wrote:
>
> Please see about |<*FATAL ANY*>| inside #791 (comment) <https://github.com/modula3/cm3/discussions/791#discussioncomment-1894374>
> and near.
>
> P.S. I continue re-reading and see good example ( I add one line):
>
> |MODULE hello2 EXPORTS Main; IMPORT IO; <* FATAL IO.Error *> <*FATAL ANY*> VAR name: TEXT; BEGIN TRY IO.Put("Gimme your name: "); name := IO.GetLine(); EXCEPT IO.Error => (*do nothing*) END; IO.Put("Your name was entered as: " & name & "\n"); END hello2.|
With Victor's addition of the TRY-EXCEPT statement,
you don't.  The signature of IO.GetLine says RAISES{Error}
to tell you that *it* can raise Error.  But the TRY-EXCEPT
wrapped around the call on GetLine will catch it and ignore
it, so it will not propagate out of this version of Hello2.

In the original version, without the TRY-EXCEPT, if the exception
came out of GetLine, it would propagate right on through hello, so
a RAISES would be appropriate.

Unfortunately, Modula-3 does not
provide a syntax for RAISES on blocks, only on procedures.  And
your code is directly inside the main block of the main module,
so there is no place.

The right way would be to pull out your code into a procedure,
put a call to that in the main block, and put the RAISES on it.



|MODULE hello2 EXPORTS Main; IMPORT IO; <*FATAL ANY*> PROCEDURE Work ( ) RAISES {IO.Error} VAR name: TEXT; BEGIN IO.Put("Gimme your name: "); name := IO.GetLine(); IO.Put("Your name was entered as: " & name & "\n"); END(*Work*) BEGIN (*hello2*) Work ( ) END hello2.|

I usually put very little in a module's block,
just a call or two on something.  If it's not the
main program module, usually nothing at all, or
a call on some module-specific but global initialization.

The only thing FATAL does is suppress the compiler warning
that a fatal exception, that is neither caught by a TRY-EXCEPT
nor announced by a RAISES, can't be ruled out by the language.
